src/pyodide-py/pyodide/console.py

```diff
@@ -134,14 +145,34 @@ def restore_stdstreams(self):
         sys.stdout = self._old_stdout
         sys.stderr = self._old_stderr
 
-    def runcode(self, code):
-        if self._persistent_stream_redirection:
-            super().runcode(code)
-        else:
+    @contextmanager
+    def stdstreams_redirections(self):
+        """ Ensure std stream redirection """
+        if not self._persistent_stream_redirection:
             self.redirect_stdstreams()
-            super().runcode(code)
+        yield
+        if not self._persistent_stream_redirection:
             self.restore_stdstreams()
 
+    def runsource(self, *args, **kwargs):
+        # syntax errors are not catched at runcode level but at runsource
+        with self.stdstreams_redirections():
+            return super().runsource(*args, **kwargs)
+
+    def runcode(self, code):
+        """ Load imported packages then run code, async. """
```

Comment:
This and and the above method would need a more detailed description in the docstring with some motivation for their existence (that you wrote in the issue description)